### PR TITLE
Correct asSuper for lower bounded wildcards

### LINKED ISSUE #1046: AsSuperVisitor: wildcards with lower bounds
The following code:

```
import java.util.List;
import java.util.function.Function;

interface EnumMarker {}

enum MyEnum implements EnumMarker {
  A,
  B;
}

class NS2Lists {
  @SuppressWarnings("nullness")
  static <F, T> List<T> transform(
      List<F> p, Function<? super F, ? extends T> q) {
    return null;
  }
}

abstract class NotSubtype2 {
  void test(List<MyEnum> p) {
    NS2Lists.transform(p, foo());
  }

  abstract Function<? super EnumMarker, Number> foo();
}
```

results in:

```
error: AsSuperVisitor: type is not an erased subtype of supertype.
  type: @Initialized @NonNull EnumMarker
  superType: @Initialized @NonNull MyEnum
  Compilation unit: NotSubtype2.java
  Exception: java.lang.Throwable; Stack trace: org.checkerframework.framework.source.SourceChecker.errorAbort(SourceChecker.java:721)
  org.checkerframework.javacutil.ErrorReporter.errorAbort(ErrorReporter.java:41)
  org.checkerframework.framework.type.AsSuperVisitor.errorTypeNotErasedSubtypeOfSuperType(AsSuperVisitor.java:155)
  org.checkerframework.framework.type.AsSuperVisitor.visitDeclared_Declared(AsSuperVisitor.java:335)
  org.checkerframework.framework.type.AsSuperVisitor.visitDeclared_Declared(AsSuperVisitor.java:29)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:308)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visit(AbstractAtmComboVisitor.java:50)
  org.checkerframework.framework.type.AsSuperVisitor.visit(AsSuperVisitor.java:86)
  org.checkerframework.framework.type.AsSuperVisitor.asSuperLowerBound(AsSuperVisitor.java:216)
  org.checkerframework.framework.type.AsSuperVisitor.asSuperWildcardLowerBound(AsSuperVisitor.java:197)
  org.checkerframework.framework.type.AsSuperVisitor.visitWildcard_Wildcard(AsSuperVisitor.java:794)
  org.checkerframework.framework.type.AsSuperVisitor.visitWildcard_Wildcard(AsSuperVisitor.java:29)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:716)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visit(AbstractAtmComboVisitor.java:50)
  org.checkerframework.framework.type.AsSuperVisitor.visit(AsSuperVisitor.java:86)
  org.checkerframework.framework.type.AsSuperVisitor.asSuper(AsSuperVisitor.java:68)
  org.checkerframework.framework.util.AnnotatedTypes.asSuper(AnnotatedTypes.java:131)
  org.checkerframework.framework.util.QualifierPolymorphism$PolyCollector.visitWildcard(QualifierPolymorphism.java:629)
  org.checkerframework.framework.util.QualifierPolymorphism$PolyCollector.visitWildcard(QualifierPolymorphism.java:336)
  org.checkerframework.framework.type.AnnotatedTypeMirror$AnnotatedWildcardType.accept(AnnotatedTypeMirror.java:1978)
  org.checkerframework.framework.type.visitor.SimpleAnnotatedTypeVisitor.visit(SimpleAnnotatedTypeVisitor.java:64)
  org.checkerframework.framework.util.QualifierPolymorphism$PolyCollector.visit(QualifierPolymorphism.java:387)
  org.checkerframework.framework.util.QualifierPolymorphism$PolyCollector.visitDeclared(QualifierPolymorphism.java:467)
  org.checkerframework.framework.util.QualifierPolymorphism$PolyCollector.visitDeclared(QualifierPolymorphism.java:336)
  org.checkerframework.framework.type.AnnotatedTypeMirror$AnnotatedDeclaredType.accept(AnnotatedTypeMirror.java:901)
  org.checkerframework.framework.type.visitor.SimpleAnnotatedTypeVisitor.visit(SimpleAnnotatedTypeVisitor.java:64)
  org.checkerframework.framework.util.QualifierPolymorphism$PolyCollector.visit(QualifierPolymorphism.java:387)
  org.checkerframework.framework.util.QualifierPolymorphism.annotate(QualifierPolymorphism.java:198)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.methodFromUse(GenericAnnotatedTypeFactory.java:1295)
  org.checkerframework.checker.nullness.NullnessAnnotatedTypeFactory.methodFromUse(NullnessAnnotatedTypeFactory.java:299)
  org.checkerframework.framework.type.TypeFromExpressionVisitor.visitMethodInvocation(TypeFromExpressionVisitor.java:171)
  org.checkerframework.framework.type.TypeFromExpressionVisitor.visitMethodInvocation(TypeFromExpressionVisitor.java:45)
  com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1477)
  com.sun.source.util.SimpleTreeVisitor.visit(SimpleTreeVisitor.java:53)
  org.checkerframework.framework.type.TypeFromTree.fromExpression(TypeFromTree.java:36)
  org.checkerframework.framework.type.AnnotatedTypeFactory.fromExpression(AnnotatedTypeFactory.java:1185)
```

Tried it with nullness, regex, and interning; probably a general issue.
A few issues with similar stack traces have been fixed. I didn't see an open issue for this.

### PR DESCRIPTION
**Pull request by smillst**
Fixes #1046

### CODE REVIEW

File: checker/tests/nullness/Issue1046.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
+// Test case for Issue 1046:
+// https://github.com/typetools/checker-framework/issues/1046
+// Addtional test case: checker/tests/nullness/java8/Issue1046.java
```
Comment (mernst) [inline checker/tests/nullness/Issue1046.java]: Should that file be part of this pull request?

Comment (smillst) [inline checker/tests/nullness/Issue1046.java]: Yes, I added it. Thanks!

File: framework/src/org/checkerframework/framework/type/AsSuperVisitor.java [hunk 2]
```diff
@@ -213,8 +213,13 @@ private AnnotatedTypeMirror asSuperLowerBound(
             lowerBound.replaceAnnotations(typeLowerBound);
             return lowerBound;
         }
-        lowerBound = visit(type, lowerBound, p);
-        return lowerBound;
+        if (areErasedJavaTypesEquivalent(type, lowerBound)) {
+            return visit(type, lowerBound, p);
+        }
+        // If type and lowerBound are not the same type, then lowerBound is a subtype of type,
+        // but there is no way to convert type to a subtype -- there is not asSub method.  So,
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/type/AsSuperVisitor.java]: Missing word "an" after "not".

**Comment by mernst** [comment]
The java8 test is failing.

**Pull request merged**